# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [15]:
#udelat opravit
def najdi_prvocisla(n):
    prvocisla = [2]
    i = prvocisla[-1]
    while len(prvocisla) < n:
        i += 1
        for prvocislo in prvocisla:
            if i % prvocislo == 0:
                break
        else:
            prvocisla.append(i)
    return prvocisla
print(najdi_prvocisla(7))



[2, 3, 5, 7, 11, 13, 17]



# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [16]:
%%writefile DU.py
def odmocnina (a, n):

    if a<0: return "Zaporne cislo"
    if a==0: return "0"
    
    b = a
    for i in range (n):
    
        b=((a/b)+b)/2
    
    return b

a = 64
print(odmocnina(a,100))


def hodnota_pi (n):
    a = 1
    b = 1


    for i in range (n):
       v = odmocnina(a-(b/2)**2,100)
       n = 6*b*v/2 * 2**i
       b = odmocnina((b/2)**2 + (1 - v)**2,100)
       
    return n
print(hodnota_pi(50))    

def aproximace_pi (n):
    odmocnina3 = odmocnina(3,1000)
    a = 1/16
    suma = a*(1/3)
    for i in range (1, n):
       a = a * ((2*(i+1)-3)/(2*(i+1)))*(1/2**2)
       suma += (1/(2*(i+1)+1))*a
    
    p = 12*(-odmocnina3/8+(1/2) - suma)

    return p

print(aproximace_pi(100))



Overwriting DU.py


In [17]:
def odmocnina (a, n):

    if a<0: return "Zaporne cislo"
    if a==0: return "0"
    
    b = a
    for i in range (n):
    
        b=((a/b)+b)/2
    
    return b
print(odmocnina(64, 100))

def hodnota_pi (n):
    a = 1
    b = 1


    for i in range (n):
       v = odmocnina(a-(b/2)**2,100)
       n = 6*b*v/2 * 2**i
       b = odmocnina((b/2)**2 + (1 - v)**2,100)
       
    return n
print(hodnota_pi(50))   




8.0
3.1415926535897922


In [18]:
%timeit odmocnina(64, 100)
%timeit hodnota_pi(50)

8.09 µs ± 52.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
746 µs ± 5.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)



# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [27]:
import numpy as np
def hodnota_pi (n):
    a = 1
    b = 1


def aproximace_pi (n):
    odmocnina3 = odmocnina(3,1000)
    a = 1/16
    suma = a*(1/3)
    for i in range (1, n):
       a = a * ((2*(i+1)-3)/(2*(i+1)))*(1/2**2)
       suma += (1/(2*(i+1)+1))*a
    
    p = 12*(-odmocnina3/8+(1/2) - suma)

    return p

print(aproximace_pi(100))
  
for n in range (1, 13):
    import math
    aprox = aproximace_pi(n)
    digit = -int(math.log10(abs(aprox - math.pi)))
    print(f"n = {n}, pi = {aprox:.5f}, počet správných cifer = {digit}")
    



3.1415926535897936
n = 1, pi = 3.15192, počet správných cifer = 1
n = 2, pi = 3.14255, počet správných cifer = 3
n = 3, pi = 3.14171, počet správných cifer = 3
n = 4, pi = 3.14161, počet správných cifer = 4
n = 5, pi = 3.14160, počet správných cifer = 5
n = 6, pi = 3.14159, počet správných cifer = 6
n = 7, pi = 3.14159, počet správných cifer = 7
n = 8, pi = 3.14159, počet správných cifer = 7
n = 9, pi = 3.14159, počet správných cifer = 8
n = 10, pi = 3.14159, počet správných cifer = 9
n = 11, pi = 3.14159, počet správných cifer = 9
n = 12, pi = 3.14159, počet správných cifer = 10



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [28]:
def Fibonacci(n):
    a = 1
    b = 1
    print (a)
    print (b)
    for i in range (n-2):
        c = a + b
        a = b
        b = c   
        print (c)
        
Fibonacci(10)

  

1
1
2
3
5
8
13
21
34
55



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [29]:
print([i for i in range(1,11)])
print([i**2 for i in range(1,11)])
print([x for x in range(2, 101) if all(x % i != 0 for i in range(2, x))]) #generuje True pro každé i menší než x, pro které platí, že x není dělitelné i. vrátí True, pokud všechny prvky v závorkách jsou True.


[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [30]:
%%writefile DU.text 
Ahoj, jak se máš?


Overwriting DU.text


In [31]:
%less DU.text



Ahoj, jak se máš?


In [32]:
%%writefile novy.text
for p in range (1,11):
  import math
  print(f"cislo = {p}, druha mocnina = {p**2}, druha odmocnina = {math.sqrt(p)}")

Overwriting novy.text


In [33]:
%run novy.text

cislo = 1, druha mocnina = 1, druha odmocnina = 1.0
cislo = 2, druha mocnina = 4, druha odmocnina = 1.4142135623730951
cislo = 3, druha mocnina = 9, druha odmocnina = 1.7320508075688772
cislo = 4, druha mocnina = 16, druha odmocnina = 2.0
cislo = 5, druha mocnina = 25, druha odmocnina = 2.23606797749979
cislo = 6, druha mocnina = 36, druha odmocnina = 2.449489742783178
cislo = 7, druha mocnina = 49, druha odmocnina = 2.6457513110645907
cislo = 8, druha mocnina = 64, druha odmocnina = 2.8284271247461903
cislo = 9, druha mocnina = 81, druha odmocnina = 3.0
cislo = 10, druha mocnina = 100, druha odmocnina = 3.1622776601683795


# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [34]:
# kostra k doplnění, místa označená TODO doplňte


class Opertor:
    def __init__(self):
        pass

    def __str__(self):
        return type(self).__name__

    def __repr__(self):
        return self.__str__()

    def diff(self):
        pass

    def __add__(self, other):
        return Addition(self, other)
        pass

    def __sub__(self, other):
        return Addition(self, Multiplication(cst(-1), other))

    def __mul__(self, other):
        return Multiplication(self, other)


class BinaryOperator(Opertor):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument

    def __str__(self):
        return f"({self.levy_argument} {type(self).__name__} {self.pravy_argument})"


class UnaryOperator(Opertor):
    def __init__(self, argument):
        self.argument = argument

    def __str__(self):
        return f"{type(self).__name__}({self.argument})"


class Addition(BinaryOperator):
    def __str__(self):
        return f"({self.levy_argument.__str__()} + {self.pravy_argument.__str__()})"
        pass

    def diff(self):
        return self.levy_argument.diff() + self.pravy_argument.diff()
        pass


class Multiplication(BinaryOperator):
    def __str__(self):
        return f"({self.levy_argument.__str__()} * {self.pravy_argument.__str__()})"
        pass

    def diff(self):
        return self.levy_argument.diff() * self.pravy_argument + self.levy_argument * self.pravy_argument.diff()
        pass

class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return str(self.value)
        pass

    def diff(self):
        return cst(0)
        pass


class identity(UnaryOperator):
    def __init__(self):
        pass

    def __str__(self):
        return "x"
        pass

    def diff(self):
        return cst(1)
        pass

class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent

    def __str__(self):
        return f"({self.argument.__str__()}^{self.exponent.__str__()})"
        pass

    def diff(self):
        return cst(self.exponent) *mocnina(self.argument,self.exponent - 1) * self.argument.diff()
        pass

class sin(UnaryOperator):
    def diff(self):
        return cos(self.argument) * self.argument.diff()
        pass


class cos(UnaryOperator):
    def diff(self):
        return cst(-1)*sin(self.argument) * self.argument.diff()
        pass

class exp(UnaryOperator):
    def diff(self):
        return exp(self.argument) * self.argument.diff()
        pass


class ln(UnaryOperator):
    def diff(self):
        return frc(self.argument)*self.argument.diff()
        pass

class frc(UnaryOperator):
    def __str__(self):
        return "1/"+self.argument.__str__()
        pass

    def diff(self):
        return cst(-1)*frc(mocnina(self.argument,2)) * self.argument.diff()

        

In [35]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

(sin((x + cos((x * x)))) + 1)
((cos((x + cos((x * x)))) * (1 + ((-1 * sin((x * x))) * ((1 * x) + (x * 1))))) + 0)


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [36]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


((((x^2) + (2 * x)) + 1) + (2 * (x^2)))
(((((2 * (x^1)) * 1) + ((0 * x) + (2 * 1))) + 0) + ((0 * (x^2)) + (2 * ((2 * (x^1)) * 1))))


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [37]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


(((x^2) + (2 * x)) + exp(x))
((((2 * (x^1)) * 1) + ((0 * x) + (2 * 1))) + (exp(x) * 1))


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...